splink/profile_data.py

```diff
@@ -117,6 +158,60 @@ def _get_df_percentiles():
     return sqls
 
 
+def _get_df_kde():
+    sql = """
+    select
+        value,
+        value_count,
+        group_name,
+    from __splink__df_all_column_value_frequencies
+    """
+    return sql
+
+
+def _get_df_correlations(column_expressions):
+    sql = f"""
+    WITH column_list AS (
+    SELECT unnest(ARRAY{column_expressions}) AS column_name
+    ),
+    column_data AS (
+    SELECT {', '.join(column_expressions)}
+    FROM __splink__df_concat
+    )
+    SELECT
+    t1.column_name AS column1,
+    t2.column_name AS column2,
+    CORR(d1.val::DOUBLE PRECISION, d2.val::DOUBLE PRECISION) AS correlation
+    FROM
+    column_list AS t1
+    CROSS JOIN
+    column_list AS t2
+    JOIN
+    (
+        SELECT
+        unnest(ARRAY{column_expressions}) AS column_name,
+        unnest(ARRAY[{', '.join(f't.{column_name}' for column_name in column_expressions)}]) AS val
+        FROM column_data AS t
+    ) AS d1 ON t1.column_name = d1.column_name
+    JOIN
+    (
+        SELECT
+        unnest(ARRAY{column_expressions}) AS column_name,
+        unnest(ARRAY[{', '.join(f't.{column_name}' for column_name in column_expressions)}]) AS val
+        FROM column_data AS t
+    ) AS d2 ON t2.column_name = d2.column_name
+    WHERE
+    t1.column_name < t2.column_name
+    GROUP BY
+    t1.column_name,
+    t2.column_name
+    ORDER BY
+    t1.column_name,
+    t2.column_name
+    """
```

Comment:
Sorry, I know it's a pain, but on this large SQL statement, can you split it up such that each CTE expression is its own SQL expression?

In Splink, we have a pipeline class that generates CTE expressions automatically when provided with raw SQL statements and table names.

This means:
1. Debug mode is more effective. Each CTE expression gets run, meaning it's much easier to identify which section of code is breaking.
2. The code is easier to read.

To summarise each each step and give you some code examples:
1. Write each individual SQL statement, adding them to a dictionary of the format [`{"sql": sql, "output_table_name": <str_name>}`](https://github.com/moj-analytical-services/splink/blob/master/splink/profile_data.py#L117).
2. Once you have these, you can queue up each SQL step using [`self._enqueue_sql`](https://github.com/moj-analytical-services/splink/blob/master/splink/linker.py#L557). This queues up a SQL statement and assigns it a CTE name.
3. After step 2, you don't need to make any additional changes. However, for completeness, see our [pipeline](https://github.com/moj-analytical-services/splink/blob/master/splink/pipeline.py) class. This processes your individual SQL statements, translating them into CTE expressions and generating your final pipeline.

Reply:
As above, code no longer needed, but thankyou for the explanation here, really helpful.